Allow user to specify LoadBalancerProfile in AzureManagedControlPlaneSpec

richardchen331 opened the pull request:
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
Allow user to specify [LoadBalancerProfile](https://github.com/Azure/azure-sdk-for-go/blob/v55.2.0/services/containerservice/mgmt/2021-05-01/containerservice/models.go#L2714) in [AzureManagedControlPlaneSpec](https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/exp/api/v1alpha4/azuremanagedcontrolplane_types.go#L28)

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #1662

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**TODOs**:
<!-- Put an "X" character inside the brackets of each completed task. Some may be optional depending on the PR. -->

- [ ] squashed commits
- [ ] includes documentation
- [ ] adds unit tests

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Allow specifying LoadBalancerProfile in AzureManagedControlPlane.
```

LochanRn commented:
/ok-to-test

alexeldeib commented on exp/api/v1alpha3/conversion_test.go:
```diff
@@ -32,17 +32,17 @@ func TestFuzzyConversion(t *testing.T) {
-		Hub:    &v1alpha4.AzureManagedCluster{},
-		Spoke:  &AzureManagedCluster{},
-	}))
+	//t.Run("for AzureMachinePool", utilconversion.FuzzTestFunc(utilconversion.FuzzTestFuncInput{
```
🙂 ?

richardchen331 replied:
hah, thanks for catching that! I must be debugging something :)

alexeldeib commented on exp/api/v1alpha3/azuremanagedcontrolplane_types.go:
```diff
@@ -81,6 +81,10 @@ type AzureManagedControlPlaneSpec struct {
+
+	// LoadBalancerProfile is the profile of the cluster load balancer.
+	// +optional
+	LoadBalancerProfile *LoadBalancerProfile `json:"loadBalancerProfile,omitempty"`
```
https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/1640#discussion_r706664686

richardchen331 replied:
Reverted change to v1a3 according to your suggestion :)

alexeldeib commented on exp/api/v1alpha4/azuremanagedcontrolplane_webhook.go:
```diff
@@ -306,3 +307,36 @@ func (r *AzureManagedControlPlane) validateSSHKey() error {
+		}
+
+		if r.Spec.LoadBalancerProfile.AllocatedOutboundPorts != nil {
+			if *r.Spec.LoadBalancerProfile.AllocatedOutboundPorts < 0 || *r.Spec.LoadBalancerProfile.AllocatedOutboundPorts > 64000 {
```
should we even allow 0?

richardchen331 replied:
According to the [api doc](https://docs.microsoft.com/en-us/rest/api/aks/managed-clusters/create-or-update#managedclusterloadbalancerprofile), the allowed values are in the range of 0 to 64000 (inclusive). Let me know if you think we shouldn't allow users to explicit specify value 0.

alexeldeib replied:
hmm, I think 0 may indicate to use the static allocation strategy. I guess that's fine 👍

alexeldeib commented on azure/services/managedclusters/managedclusters.go:
```diff
@@ -58,6 +58,14 @@ type Service struct {
 	Client
 }
 
+func convertToResouceReferences(resources []string) *[]containerservice.ResourceReference {
```
```suggestion
func convertToResourceReferences(resources []string) *[]containerservice.ResourceReference {
```

alexeldeib commented on azure/services/managedclusters/managedclusters.go:
```diff
@@ -173,6 +181,21 @@ func (s *Service) Reconcile(ctx context.Context) error {
+				PublicIPs: convertToResouceReferences(managedClusterSpec.LoadBalancerProfile.OutboundIPs),
+			},
+			EffectiveOutboundIPs:   convertToResouceReferences(managedClusterSpec.LoadBalancerProfile.EffectiveOutboundIPs),
+			AllocatedOutboundPorts: managedClusterSpec.LoadBalancerProfile.ManagedOutboundIPs,
+			IdleTimeoutInMinutes:   managedClusterSpec.LoadBalancerProfile.ManagedOutboundIPs,
```
```suggestion
			AllocatedOutboundPorts: managedClusterSpec.LoadBalancerProfile.ManagedOutboundIPs,
			IdleTimeoutInMinutes:   managedClusterSpec.LoadBalancerProfile.ManagedOutboundIPs,
```
```suggestion
			AllocatedOutboundPorts: managedClusterSpec.LoadBalancerProfile.AllocatedOutboundPorts,
			IdleTimeoutInMinutes:   managedClusterSpec.LoadBalancerProfile.IdleTimeoutInMinutes,
```

richardchen331 replied:
Thanks for catching these mistakes!

alexeldeib commented on azure/types.go:
```diff
@@ -372,6 +375,27 @@ type AADProfile struct {
+// LoadBalancerProfile - Profile of the cluster load balancer.
+type LoadBalancerProfile struct {
+	// ManagedOutboundIPs - Desired managed outbound IPs for the cluster load balancer.
+	ManagedOutboundIPs *int32
```
nit: a slightly clearer description of why you'd use this vs outboundIPs might be nice. It's fairly clear from the API one is managed by AKS and one is BYO, but would be nice to be explicit.

alexeldeib commented on exp/api/v1alpha4/azuremanagedcontrolplane_types.go:
```diff
@@ -100,6 +104,33 @@ type AADProfile struct {
+
+	// EffectiveOutboundIPs - The effective outbound IP resources of the cluster load balancer.
+	// +optional
+	EffectiveOutboundIPs []string `json:"effectiveOutboundIPs,omitempty"`
```
are you familiar with the impact of sending this value? 

I believe it's a ready only field which is the union of the resource IDs from OutboundIPs and the resource IDs of the AKS-created ManagedOutboundIPs, and thus should not be in our spec.

richardchen331 replied:
Thanks for pointing it out! I checked the [api doc](https://docs.microsoft.com/en-us/rest/api/aks/managed-clusters/create-or-update) again and it's indeed only in the api response, i'll remove it from the spec.

richardchen331 commented:
Removed `EffectiveOutboundIPs` from spec. Added additional validation logic.

alexeldeib commented:
/approve

nice work!

richardchen331 commented:
> /approve
> 
> nice work!

thanks for reviewing all my PRs!

richardchen331 commented:
/retest

CecileRobertMichon commented:
/assign @LochanRn

LochanRn commented:
@richardchen331 can you please update the documentation for this as well :) 

Nice work ❤️. 

@CecileRobertMichon  PR lgtm.

richardchen331 commented:
> @richardchen331 can you please update the documentation for this as well :)
> 
> Nice work ❤️.
> 
> @CecileRobertMichon PR lgtm.

Thanks! Sure, just updated the documentation for this feature :)

CecileRobertMichon commented on docs/book/src/topics/managedcluster.md:
```diff
@@ -102,6 +102,15 @@ spec:
+  loadBalancerProfile: # Load balancer profile must specify at most one of ManagedOutboundIPs, OutboundIPPrefixes and OutboundIPs
+    managedOutboundIPs: 2 # 1-100
+    outboundIPPrefixes:
+    - /subscriptions/fae7cc14-bfba-4471-9435-f945b42a16dd/resourceGroups/foo-bar/providers/Microsoft.Network/publicIPPrefixes/my-public-ip-prefix # fake public ip prefix
```
@richardchen331 can you please remove the sub ID from the example? replace it by something that is obviously fake like `123`

richardchen331 replied:
sure, replaced it with `00000000-0000-0000-0000-000000000000` :)

LochanRn commented on docs/book/src/topics/managedcluster.md:
```diff
@@ -98,17 +98,26 @@ spec:
   version: v1.21.2
   networkPolicy: azure # or calico
   networkPlugin: azure # or kubenet
+  sku: Free # or Paid
+  loadBalancerProfile: # Load balancer profile must specify at most one of ManagedOutboundIPs, OutboundIPPrefixes and OutboundIPs
+    managedOutboundIPs: 2 # 1-100
+    outboundIPPrefixes:
+    - /subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/foo-bar/providers/Microsoft.Network/publicIPPrefixes/my-public-ip-prefix # fake public ip prefix
+    outboundIPs:
+    - /subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/foo-bar/providers/Microsoft.Network/publicIPAddresses/my-public-ip # fake public ip
+    allocatedOutboundPorts: 100 # 0-64000
+    idleTimeoutInMinutes: 10 # 4-120
```
Because these fields are optional can we put this in a separate section just like mutitenancy and aad profile. 
The reason why I am saying this is, not all users might want to specify a separate load balancer

Also when you add in the section please link the original loadBalancerProfile docs and if possible a small brief about it.

LochanRn commented:
@richardchen331 could you please add a few test cases for azuremanagedcontrolplane_webhook for loadbalancerprofile validator !!

richardchen331 commented:
> @richardchen331 sorry for the delayed request, should have asked it the first time, but i missed it, could you please add a few unit tests for loadbalancerprofile validator !!

Thanks @LochanRn for your review! I've updated the doc according to your suggestion and added unit tests as well.

LochanRn commented:
@richardchen331 Can you please squash all your commits !!

LochanRn commented:
Thanks again for all your work, the PR is good to go 👍

richardchen331 commented:
> Thanks again for all your work, the PR is good to go 👍

Thanks for reviewing it! Just squashed all commits :)

richardchen331 commented:
/retest

richardchen331 commented:
/retest

LochanRn commented:
/test pull-cluster-api-provider-azure-e2e-exp

LochanRn commented:
/test pull-cluster-api-provider-azure-e2e-exp

LochanRn commented:
@richardchen331  can you please run the test-case locally and check if everything is fine. Thank you.

CecileRobertMichon commented:
@alexeldeib @LochanRn the AKS test has been very flaky over the last week https://testgrid.k8s.io/sig-cluster-lifecycle-cluster-api-provider-azure#capz-periodic-capi-e2e-full-main

richardchen331 commented:
/retest

richardchen331 commented:
@LochanRn @CecileRobertMichon @alexeldeib I ran the failed e2e test locally and it passed, so I'll rerun the test now. 

I think the flakiness is probably related to the issue mentioned in the slack channel (https://kubernetes.slack.com/archives/CEX9HENG7/p1631738756123200). We do see the issue (system node pool stuck in Updating state) very often when we provision an AKS cluster with CAPZ.

LochanRn commented:
/lgtm

richardchen331 commented:
/retest

LochanRn commented:
@CecileRobertMichon will take a look at the e2e tests over the weekend !!

richardchen331 commented:
/retest
